### PR TITLE
feat(CloudPakForDataServiceInstanceAuthenticator): add new cp4d service instance authenticator

### PR DESCRIPTION
**Pull request by jpfloresibm**
This commit introduces a new authenticator implementation that supports the "GET /v3/service_instances/serviceInstanceId/token"
operation which is part of the Cloud Pak For Data authentication service. This allows users to obtain
access tokens for service instances using the username, apikey, and service instance id in CP4D.

### CODE REVIEW

File: src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorLiveTest.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+// CP4DTEST1_AUTH_DISABLE_SSL=true
+//
+// Then remove/comment-out the @Ignore annotation below and run the method as a junit test.
+//
+public class Cp4dServiceInstanceAuthenticatorLiveTest {
+
+  @Ignore
+  @Test
+  public void testCp4dLiveTokenServer() {
+	System.setProperty("IBM_CREDENTIALS_FILE", "cp4dtest.env");
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorLiveTest.java:50]:
```suggestion
	System.setProperty("IBM_CREDENTIALS_FILE", "cp4d_service_instance.env");
```
Using a different filename like this will let us keep the config for this testcase separate from the "live test" of the original CloudPakForDataAuthenticator, which also uses this file.   Note that you used the filename `cp4dserivicetest.env` in the CloudPakForDataServiceAuthenticator.

File: src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    url = getMockWebServerUrl();
+    tokenData = loadFixture("src/test/resources/cp4d_service_instance_token.json",
+        Cp4dServiceInstanceTokenResponse.class);
+    refreshedTokenData = loadFixture("src/test/resources/refreshed_cp4d_service_instance_token.json",
+        Cp4dServiceInstanceTokenResponse.class);
+  }
+
+
+  //
+  // Tests involving the new Builder class and fromConfiguration() method.
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java:78]:
```suggestion
  // Tests involving the Builder class and fromConfiguration() method.
```
The whole thing is new at this point :)

File: src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+  @Test(expectedExceptions = IllegalArgumentException.class)
+  public void testBuilderEmptyApikey() {
+    new CloudPakForDataServiceInstanceAuthenticator.Builder()
+      .url("https://good-url")
+      .username(testUsername)
+      .apikey("")
+      .serviceInstanceId(testServiceInstanceId)
+      .build();
+  }
+
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java:139]: Also need tests for a missing/empty service_instance_id value.

File: src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+  @Test(expectedExceptions = IllegalArgumentException.class)
+  public void testConfigMissingApikey() {
+    Map<String, String> props = new HashMap<>();
+    props.put(Authenticator.PROPNAME_URL, "https://good-url");
+    props.put(Authenticator.PROPNAME_USERNAME, testUsername);
+    props.put(Authenticator.PROPNAME_APIKEY, null);
+    props.put(Authenticator.PROPNAME_SERVICE_INSTANCE_ID, testServiceInstanceId);
+    CloudPakForDataAuthenticator.fromConfiguration(props);
+  }
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java:167]: Also need a test for a missing service_instance_id value.

File: src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+      assertNotNull(causedBy);
+      assertTrue(causedBy instanceof IllegalStateException);
+    } catch (Throwable t) {
+      fail("Expected RuntimeException, not " + t.getClass().getSimpleName());
+    }
+  }
+
+  @Test(expectedExceptions =IllegalArgumentException.class)
+  public void testCtorMissingUsernameMap() {
+    Map<String, String> props = new HashMap<>();
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java:454]: it looks like this test and the rest of the tests in this class are duplicates of testcases in the first section of test methods.

**Review by padamstx — CHANGES_REQUESTED** [review]
Outstanding job on this PR!   In addition to the things we discussed on webex, I requested a few small changes that are mostly just cosmetic, plus I think you have a handful of duplicated testcases (probably due to the fact that other authenticators need to continue to test the legacy ctor, etc.).

**Review by padamstx — APPROVED** [review]
I've reviewed the recent changes to the PR and it looks like you've addressed all my comments.
Great job on this PR!
